### ISSUE TITLE
Some queries slower when enabling the Interpreter.

### ISSUE THREAD

**Issue by RynoPi** [rmosolgo/graphql-ruby#3345]
When we enable the interpreter we realize that some of our queries run slower than before while others run faster. Our assumption was that enabling the interpreter should either speed up or stay the same. Clearly some configuration our setup in some of our types are causing this, but cannot seem to find anything on the internet that would indicate why.

We also tried upgrading to the latest version (which has it enabled by default), but still experience them running slower than disabling the interpreter. We noticed that most of them that run slower have a scope applied so it might be the reason we are having this experience.

Would using a Resolver have worse performance when using the runtime interpreter?

If there is any reason that might indicate why, it would be much appreciated.

**Comment by rmosolgo** [comment]
Hi! It's certainly possible that something might be slower, although I don't know what. In order to debug this, I'll need some specifics: 

- An example GraphQL query string that runs slower on the interpreter than the legacy runtime 
- A benchmark of the query on the interpreter, for example: 

  ```ruby 
   StackProf.run(mode: :wall, out: "interpreter-profile.dump") do 
     MySchema.execute(...)
   end 
   ```
   
   Then share the `.dump` file or at least the output of `stackprof interpreter-profile.dump`
- If certain GraphQL fields or resolvers are identified as obviously the slowest, then sharing the source code of those would be really helpful. 


After you share some of that info, we can dig in!

**Comment by RynoPi** [comment]
Sorry for taking so long to reply. Here is the top 30 with the legacy runtime:
![Screenshot from 2021-02-18 14-39-49](https://user-images.githubusercontent.com/77432729/108358269-31b8fd80-71f7-11eb-904a-13193a84b76a.png)
 
Here is the top 30 with the new interpreter from StackProf:
![Screenshot from 2021-02-18 14-12-17](https://user-images.githubusercontent.com/77432729/108355640-c7eb2480-71f3-11eb-8a95-f2b223a98e79.png)

Regarding the Graphql query I obfuscated some of the data. You will see a couple of fields name `aField` these all represent a different field. Here is the Graphql query:
```
query OrdersSearch($limit: Int!, $skip: Int!, $filter: Filter!, $itemstates: [String!], $weekstart: String, $weekend: String) {
    ordersSearch(filter: $filter) {
      orders(limit: $limit, skip: $skip) {
        ...multipleOrderFields
        __typename
      }
      totalOrders
      totalFiltered
      __typename
    }
  }

  fragment multipleOrderFields on Order {
    id
    state
    fieldA
    updatedAt
    createdAt
    itemGroupings {
      id
      cumulativeQuantity
      quantityType
      commonFields
      deleted
      rank
      items(itemStates: $itemstates, weekStart: $weekstart, weekEnd: $weekend) {
        ...singleItemFields
        ...otherFields
        __typename
      }
      __typename
    }
    buyer {
      id
      name
      logoFile {
        id
        filename
        url
        __typename
      }
      __typename
    }
    seller {
      id
      name
      logoFile {
        id
        filename
        url
        __typename
      }
      categories {
        id
        categoryId
        name
        productType {
          id
          name
          __typename
        }
        __typename
      }
      __typename
    }
    __typename
  }

  fragment singleItemFields on Item {
    id
    additionalFields
    fullWeek
    startWeek
    endWeek
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    aField
    categoryId
    varietyId
    packedCartons
    packedPallets
    product {
      id
      category {
        id
        name
        __typename
      }
      __typename
    }
    referenceNumber
    rank
    __typename
  }

  fragment otherFields on Item {
    other {
      aField
      aField
      aField
      aField
      aField
      aField
      __typename
    }
    __typename
  }
```
Please let me know if any of this is not helpful.

**Comment by rmosolgo** [comment]
Thanks for sharing those details, that's really helpful. I was hoping to spot a smoking gun in the stackprof output, but there really isn't anything. It looks like there's just _more_ GraphQL overhead involved in running this query. 

I'd like to know what factor slowdown we're looking at here. Could you share some high-level benchmarks for this query? For example, running a benchmark like this _with_ and _without_ the new runtime: 

```ruby
# bundle add benchmark-ips && bundle install 
Benchmark.ips do |x| 
  x.report("Query") { MySchema.execute(...) }
end 
```

It might be possible to eliminate some of the overhead shown in the stackprof output above, but I'd like to know how much overall difference there is.

**Comment by RynoPi** [comment]
Thanks for gem. The responses should be the same ( it is consistent when I switch between the old and the new runtime).

We have the current setup to do some blanket testing on queries created by our front-end
* Store each query in a file
* Loop over all the files and run the query through Graphql.execute

I modified the loop with the Benchmark code to store the benchmarks for each query in two files for the old and new runtime:
* [Benchmarks.zip](https://github.com/rmosolgo/graphql-ruby/files/6009104/Benchmarks.zip)

Query_twelve is the example mentioned above. If I read that correctly, it went from 3 iteration to 2 iteration per 5 seconds. That would indicate a 33% decrease.

**Comment by rmosolgo** [comment]
Thanks for sharing that. I'd love to understand the format. Here's what I found in the zip for `query_twelve`: 

```json
# old.json 
{
  "item": "query_twelve",
  "measured_us": 7404240.416000366,
  "iter": 3,
  "samples": [0.4145795945628497, 0.4077589021202594, 0.39374283340065264],
  "cycles": 1
}
```

```json 
# new.json 
{
  "item": "query_twelve",
  "measured_us": 6111308.505981445,
  "iter": 2,
  "samples": [0.3309883773293423, 0.3236188999872797],
  "cycles": 1
}
```

What do the different pairs in that object represent?

**Comment by rmosolgo** [comment]
👋  Just following up on this because I'd love to get to the bottom of it. Could you please share some details about the profiling artifacts you shared? I'd love to know: 

> What do the different pairs in that object represent? (Or, could you share the code that populates those benchmarks?)

**Comment by RynoPi** [comment]
Apologies. I got caught up with some work that requires my attention (almost done with it).

I am glad I am not the only person confused by the output. I could not find anything yet that indicates what each property means. I saw in the code that `us` refers to time in milliseconds so I guessed that means that `measured_us` and `iter` refers to how long it took to complete the number of iterations. Since I am not 100% sure I want to go through the setup again and see if I can get something more concrete.

Regarding the samples, I have no idea. Ill try to get a better example tomorrow.

**Comment by rmosolgo** [comment]
> I could not find anything yet that indicates what each property means

Oh! I didn't recognize the output, so I assumed it was an internal tool or script. Do you have access to the code that generates that output?

**Comment by RynoPi** [comment]
Here is a snippet of the rake task that we wrote to test and generate the benchmarks:
``` ruby
Benchmark.ips do |x|
  x.save! "new.json"
  Dir.each_child('./lib/apq') do |filename|
    apq = YAML.load_file("./lib/apq/#{filename}")
    apq.deep_symbolize_keys!
    query = <<-GRAPHQL
      #{apq[:query]}
    GRAPHQL

    x.report(filename.sub('.yml', '')) do
      MySchema.execute(query, context: context, variables: apq[:variables])
    end
  end
end
```
We ran it once with the old runtime and replaced `"new.json"` with `"old.json"` and then ran it again with the new runtime saving it as `"new.json"`.

**Comment by RynoPi** [comment]
I got the benchmarks to work. 
```
{"version"=>"old", "method"=>"q1"}:      671.5 i/s
{"version"=>"new", "method"=>"q1"}:      612.6 i/s - same-ish: difference falls within error
{"version"=>"old", "method"=>"q2"}:      312.1 i/s - 2.15x  (± 0.00) slower
{"version"=>"new", "method"=>"q2"}:      296.1 i/s - 2.27x  (± 0.00) slower
{"version"=>"new", "method"=>"q3"}:      207.8 i/s - 3.23x  (± 0.00) slower
{"version"=>"old", "method"=>"q3"}:      194.6 i/s - 3.45x  (± 0.00) slower
{"version"=>"old", "method"=>"q4"}:      140.8 i/s - 4.77x  (± 0.00) slower
{"version"=>"new", "method"=>"q5"}:      138.4 i/s - 4.85x  (± 0.00) slower
{"version"=>"new", "method"=>"q4"}:      134.9 i/s - 4.98x  (± 0.00) slower
{"version"=>"old", "method"=>"q5"}:      117.8 i/s - 5.70x  (± 0.00) slower
{"version"=>"old", "method"=>"q6"}:       81.1 i/s - 8.28x  (± 0.00) slower
{"version"=>"old", "method"=>"q7"}:       79.6 i/s - 8.44x  (± 0.00) slower
{"version"=>"new", "method"=>"q6"}:       77.7 i/s - 8.64x  (± 0.00) slower
{"version"=>"new", "method"=>"q7"}:       75.5 i/s - 8.89x  (± 0.00) slower
{"version"=>"new", "method"=>"q8"}:       48.2 i/s - 13.92x  (± 0.00) slower
{"version"=>"old", "method"=>"q8"}:       47.7 i/s - 14.08x  (± 0.00) slower
{"version"=>"new", "method"=>"q9"}:       31.5 i/s - 21.29x  (± 0.00) slower
{"version"=>"old", "method"=>"q9"}:       31.5 i/s - 21.32x  (± 0.00) slower
{"version"=>"new", "method"=>"q10"}:       28.1 i/s - 23.87x  (± 0.00) slower
{"version"=>"old", "method"=>"q10"}:       27.5 i/s - 24.39x  (± 0.00) slower
{"version"=>"new", "method"=>"q11"}:       17.8 i/s - 37.79x  (± 0.00) slower
{"version"=>"new", "method"=>"q12"}:       16.4 i/s - 40.84x  (± 0.00) slower
{"version"=>"old", "method"=>"q11"}:       16.0 i/s - 42.01x  (± 0.00) slower
{"version"=>"new", "method"=>"q13"}:       15.8 i/s - 42.46x  (± 0.00) slower
{"version"=>"new", "method"=>"q14"}:       13.6 i/s - 49.41x  (± 0.00) slower
{"version"=>"old", "method"=>"q13"}:       13.5 i/s - 49.70x  (± 0.00) slower
{"version"=>"old", "method"=>"q12"}:       13.3 i/s - 50.66x  (± 0.00) slower
{"version"=>"old", "method"=>"q14"}:       11.8 i/s - 56.70x  (± 0.00) slower
{"version"=>"new", "method"=>"q15"}:       11.5 i/s - 58.43x  (± 0.00) slower
{"version"=>"old", "method"=>"q15"}:       11.3 i/s - 59.21x  (± 0.00) slower
{"version"=>"new", "method"=>"q16"}:        9.6 i/s - 69.92x  (± 0.00) slower
{"version"=>"new", "method"=>"q17"}:        8.9 i/s - 75.76x  (± 0.00) slower
{"version"=>"old", "method"=>"q16"}:        8.8 i/s - 76.65x  (± 0.00) slower
{"version"=>"old", "method"=>"q17"}:        7.9 i/s - 84.70x  (± 0.00) slower
{"version"=>"new", "method"=>"q18"}:        3.7 i/s - 179.39x  (± 0.00) slower
{"version"=>"new", "method"=>"q19"}:        3.3 i/s - 205.94x  (± 0.00) slower
{"version"=>"new", "method"=>"q20"}:        3.0 i/s - 226.64x  (± 0.00) slower
{"version"=>"old", "method"=>"q20"}:        2.8 i/s - 242.76x  (± 0.00) slower
{"version"=>"old", "method"=>"q18"}:        2.7 i/s - 244.54x  (± 0.00) slower
{"version"=>"old", "method"=>"q19"}:        2.4 i/s - 285.54x  (± 0.00) slower
{"version"=>"new", "method"=>"q21"}:        0.4 i/s - 1613.44x  (± 0.00) slower
{"version"=>"old", "method"=>"q21"}:        0.3 i/s - 2030.64x  (± 0.00) slower
{"version"=>"new", "method"=>"q22"}:        0.0 i/s - 18034.65x  (± 0.00) slower
{"version"=>"old", "method"=>"q22"}:        0.0 i/s - 23400.29x  (± 0.00) slower
```
I leave this here for now, until I can figure out how to properly group them

**Comment by rmosolgo** [comment]
Thanks for sharing those details. I never knew about Benchmark-ips's `save!` method, neat! 

I parsed that code and used Google sheets to pivot and compare the results: 


query | new i/s | old i/s | % slowdown
-- | -- | -- | --
q22 | 0 | 0 | 0
q19 | 2.3 | 3.2 | 28.13%
q21 | 0.3 | 0.4 | 25.00%
q18 | 2.8 | 3.7 | 24.32%
q12 | 13.5 | 16.3 | 17.18%
q14 | 13.2 | 15.8 | 16.46%
q5 | 116.8 | 135.2 | 13.61%
q13 | 11.7 | 13.4 | 12.69%
q16 | 8.5 | 9.4 | 9.57%
q11 | 15.8 | 17.4 | 9.20%
q17 | 7.8 | 8.5 | 8.24%
q3 | 192.4 | 204.6 | 5.96%
q10 | 27 | 28.1 | 3.91%
q20 | 2.8 | 2.9 | 3.45%
q8 | 47.1 | 48 | 1.88%
q15 | 11 | 11 | 0.00%
q9 | 31.1 | 31 | -0.32%
q6 | 79.5 | 77.2 | -2.98%
q7 | 77.8 | 74.6 | -4.29%
q2 | 311.3 | 295.5 | -5.35%
q4 | 139.5 | 132.3 | -5.44%
q1 | 656 | 601.5 | -9.06%


I'll take a look at reducing the library overhead highlighted in the stackprof you shared above, and follow up here when I have something. Thanks again for digging in on this!

**Comment by rmosolgo** [comment]
I've looked into a few methods over at #3364, could you give your benchmarks (and stackprof) another try with that branch? You can install it with: 

```ruby 
gem "graphql", github: "rmosolgo/graphql-ruby", branch: "optimize-interpreter-overhead"
```

Let me know what you find!

**Comment by RynoPi** [comment]
Here is the new benchmarks results (unfortunately seems to be mostly the same):
```
Comparison:
{"version"=>"new", "method"=>"q1"}:      617.6 i/s
{"version"=>"old", "method"=>"q1"}:      594.1 i/s - same-ish: difference falls within error
{"version"=>"new", "method"=>"q2"}:      304.6 i/s - 2.03x  (± 0.00) slower
{"version"=>"old", "method"=>"q2"}:      287.5 i/s - 2.15x  (± 0.00) slower
{"version"=>"old", "method"=>"q3"}:      200.2 i/s - 3.08x  (± 0.00) slower
{"version"=>"new", "method"=>"q3"}:      191.6 i/s - 3.22x  (± 0.00) slower
{"version"=>"new", "method"=>"q4"}:      134.0 i/s - 4.61x  (± 0.00) slower
{"version"=>"old", "method"=>"q5"}:      132.6 i/s - 4.66x  (± 0.00) slower
{"version"=>"old", "method"=>"q4"}:      130.1 i/s - 4.75x  (± 0.00) slower
{"version"=>"new", "method"=>"q5"}:      109.0 i/s - 5.67x  (± 0.00) slower
{"version"=>"new", "method"=>"q6"}:       80.5 i/s - 7.67x  (± 0.00) slower
{"version"=>"old", "method"=>"q7"}:       75.3 i/s - 8.20x  (± 0.00) slower
{"version"=>"old", "method"=>"q6"}:       72.8 i/s - 8.48x  (± 0.00) slower
{"version"=>"new", "method"=>"q7"}:       72.3 i/s - 8.54x  (± 0.00) slower
{"version"=>"old", "method"=>"q8"}:       46.9 i/s - 13.18x  (± 0.00) slower
{"version"=>"new", "method"=>"q8"}:       46.6 i/s - 13.25x  (± 0.00) slower
{"version"=>"new", "method"=>"q9"}:       30.4 i/s - 20.34x  (± 0.00) slower
{"version"=>"old", "method"=>"q9"}:       30.1 i/s - 20.49x  (± 0.00) slower
{"version"=>"new", "method"=>"q10"}:       26.5 i/s - 23.33x  (± 0.00) slower
{"version"=>"old", "method"=>"q10"}:       26.5 i/s - 23.34x  (± 0.00) slower
{"version"=>"old", "method"=>"q11"}:       16.9 i/s - 36.60x  (± 0.00) slower
{"version"=>"old", "method"=>"q12"}:       15.7 i/s - 39.40x  (± 0.00) slower
{"version"=>"new", "method"=>"q11"}:       15.5 i/s - 39.88x  (± 0.00) slower
{"version"=>"old", "method"=>"q14"}:       15.5 i/s - 39.95x  (± 0.00) slower
{"version"=>"new", "method"=>"q12"}:       14.2 i/s - 43.56x  (± 0.00) slower
{"version"=>"old", "method"=>"q13"}:       13.0 i/s - 47.58x  (± 0.00) slower
{"version"=>"new", "method"=>"q14"}:       12.0 i/s - 51.34x  (± 0.00) slower
{"version"=>"new", "method"=>"q13"}:       11.7 i/s - 52.93x  (± 0.00) slower
{"version"=>"new", "method"=>"q15"}:       10.9 i/s - 56.84x  (± 0.00) slower
{"version"=>"old", "method"=>"q15"}:       10.8 i/s - 57.27x  (± 0.00) slower
{"version"=>"old", "method"=>"q16"}:        9.3 i/s - 66.75x  (± 0.00) slower
{"version"=>"old", "method"=>"q16"}:        8.2 i/s - 74.90x  (± 0.00) slower
{"version"=>"new", "method"=>"q17"}:        8.2 i/s - 75.10x  (± 0.00) slower
{"version"=>"new", "method"=>"q17"}:        7.6 i/s - 81.52x  (± 0.00) slower
{"version"=>"old", "method"=>"q18"}:        3.5 i/s - 177.19x  (± 0.00) slower
{"version"=>"old", "method"=>"q19"}:        3.2 i/s - 195.20x  (± 0.00) slower
{"version"=>"new", "method"=>"q20"}:        2.9 i/s - 214.95x  (± 0.00) slower
{"version"=>"old", "method"=>"q18"}:        2.8 i/s - 218.34x  (± 0.00) slower
{"version"=>"new", "method"=>"q20"}:        2.7 i/s - 226.68x  (± 0.00) slower
{"version"=>"new", "method"=>"q19"}:        2.3 i/s - 264.33x  (± 0.00) slower
{"version"=>"old", "method"=>"q21"}:        0.4 i/s - 1539.65x  (± 0.00) slower
{"version"=>"new", "method"=>"q21"}:        0.3 i/s - 1867.15x  (± 0.00) slower
{"version"=>"old", "method"=>"q22"}:        0.0 i/s - 17139.75x  (± 0.00) slower
{"version"=>"new", "method"=>"q22"}:        0.0 i/s - 20405.95x  (± 0.00) slower
```
I used q19 to get a StackProf, because it is the same query mentioned at the top of this thread.

Here is the StackProf on the old runtime:
![Screenshot from 2021-02-25 11-56-22](https://user-images.githubusercontent.com/77432729/109136462-cfb14880-7760-11eb-9928-c924dc320129.png)

Here is the StackProf on the new runtime:
![Screenshot from 2021-02-25 11-58-16](https://user-images.githubusercontent.com/77432729/109136475-d50e9300-7760-11eb-86da-534a9e4139f1.png)

**Comment by RynoPi** [comment]
Not sure if the issue is related since it might indicate an incorrect setup, we are running into the following error `Pundit::NotDefinedError - unable to find scope `GraphQL::Execution::Execute::SkipPolicy::Scope` with the following code:
``` Ruby
# frozen_string_literal: true

module Types
  class SubscriptionType < Types::BaseObject
    extend GraphQL::Subscriptions::SubscriptionRoot
    field :new_user_feed, Types::UserFeedIType, null: false, subscription_scope: :current_user_id
    field :commands, [Types::CommandType], null: false, subscription_scope: :current_user_id
  end
end
```
We were able to get past the error by defining the field methods as follow:
``` Ruby
def new_user_feed
  object
end

def commands
  object
end
```

**Comment by RynoPi** [comment]
Appreciate your help. [Here](https://github.com/rmosolgo/graphql-ruby/files/6117737/q19_memory_profiler.txt) is the output for q19 using the _new_ runtime. Let me know if there is anything else I could help with.

**Comment by rmosolgo** [comment]
Thanks so much for sharing that. I'm surprised to see a lot of allocations unrelated to GraphQL:

```
allocated objects by location
-----------------------------------
     79359  /usr/local/bundle/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23
     55118  /usr/local/bundle/gems/binding_of_caller-1.0.0/lib/binding_of_caller/mri.rb:21
     32746  /usr/local/bundle/gems/better_errors-2.9.1/lib/better_errors/exception_extension.rb:6
     22135  /usr/local/bundle/gems/listen-3.1.5/lib/listen/record/entry.rb:53
     17349  /usr/local/bundle/bundler/gems/graphql-ruby-4902fe31af0a/lib/graphql/execution/interpreter/runtime.rb:181
     17349  /usr/local/bundle/bundler/gems/graphql-ruby-4902fe31af0a/lib/graphql/execution/interpreter/runtime.rb:271
```

But, this might have something to do with Rails autoloading (seeing Bootsnap on the list). 

What if you modify the script to include a "warm up run" of the query? That way we could rule out any autoloading-related overhead, for example: 

```ruby 
require 'memory_profiler'

# first, run the query once here 

report = MemoryProfiler.report do
  # then, run the query again here 
end

report.pretty_print(to_file: "graphql_query_memory.txt")
```

After the modification, if the memory profile has the same top objects by location, we can investigate those specific locations to see where they're coming from. If the memory profile is _different_ ... then we'll go from there!

**Comment by RynoPi** [comment]
[Here](https://github.com/rmosolgo/graphql-ruby/files/6121060/q19_memory_profiler.txt) is the new results as per your recommendation. It looks like the result is different.

**Comment by RynoPi** [comment]
We realized that our results might not be 100% correct since we have been running these performance tests in a dev environment.

Here is the Benchmark results running them in a production like environment for query q19.

Comparison:
{"version"=>"old", "method"=>"q19"}:        1.6 i/s
{"version"=>"new", "method"=>"q19"}:        1.2 i/s - 1.30x  (± 0.00) slower

[Here](https://github.com/rmosolgo/graphql-ruby/files/6128349/new_memory.txt) is the memory profile report as well for this environment.

**Comment by syossan27** [comment]
I was having the same problem.
I don't know what caused it, but I changed it to run the query without variables and that improved the performance degradation.
I hope this is of some help.

**Comment by rmosolgo** [comment]
That's very interesting, @syossan27 , thanks for sharing! I'll try adding variables to some of my local benchmarks and see if I can replicate the problem better.

**Comment by rmosolgo** [comment]
I just shipped 1.12.11 which included a pretty big performance improvement (#3494). If you get a chance, I'd love to see your benchmark results on that new version. 

Besides that, I don't have plans to look for more performance improvements, so I'm going to close this. We can reopen it if you don't see an improvement on that version.
